src/org/omegat/core/data/ExternalTMFactory.java

```diff
@@ -118,26 +123,45 @@ public ExternalTMX load(Language sourceLang, Language targetLang) throws Excepti
             TMXReader2.LoadCallback loader = new TMXReader2.LoadCallback() {
                 public boolean onEntry(TMXReader2.ParsedTu tu, TMXReader2.ParsedTuv tuvSource,
                         TMXReader2.ParsedTuv tuvTarget, boolean isParagraphSegtype) {
+
                     if (tuvSource == null) {
                         return false;
                     }
 
+                    // Keep all the Tuvs matching at least the target language
+                    for (TMXReader2.ParsedTuv tuv : tu.tuvs) {
+                        String[] tuvLangParts = tuv.lang.split("\\W");
+                        if (!targetLang.getLanguageCode().equalsIgnoreCase(tuvLangParts[0])) {
+                            continue;
+                        }
+                        addTuv(tu, tuvSource, tuv, isParagraphSegtype);
+                    }
+
+                    // If there was no match, use the original tuvTarget
                     if (tuvTarget != null) {
-                        // add only target Tuv
-                        addTuv(tu, tuvSource, tuvTarget, isParagraphSegtype);
+                        if (entries.isEmpty()) {
```

Comment:
As I understand it, this check is to make sure we don't add a duplicate in case we already added the target in the `// Keep all the Tuvs matching at least the target language` block above.

This makes me nervous, because we are reaching "below" the current "layer" and looking at something that *coincidentally* may tell us the right answer, but is not guaranteed to. I would much rather fix the logic of this method to call `addTuv` the right number of times without looking at `entries`.

Reply:
I rewrote the matching loops to be a bit clearer and keep/leave out matches only as needed.